Sources/MapboxCoreNavigation/BillingHandler.swift

```diff
@@ -0,0 +1,379 @@
+            onError(BillingServiceError(nativeBillingServiceError))
+        }
+    }
+
+    func stopBillingSession(for sessionType: BillingHandler.SessionType) {
+        let skuToken = tripSku(for: sessionType)
+        BillingServiceNative.stopBillingSession(for: skuToken)
+    }
+
+    func triggerBillingEvent(onError: @escaping (BillingServiceError) -> Void) {
+        BillingServiceNative.triggerBillingEvent(forAccessToken: accessToken,
+                                                 userAgent: userAgent,
+                                                 skuIdentifier: mauSku) { nativeBillingServiceError in
+            onError(BillingServiceError(nativeBillingServiceError))
+        }
+    }
+
+    func getSessionStatus(for sessionType: BillingHandler.SessionType) -> BillingHandler.SessionState {
+        switch BillingServiceNative.getSessionStatus(for: tripSku(for: sessionType)) {
+        case .noSession: return .stopped
+        case .sessionActive: return .running
+        case .sessionPaused: return .paused
+        @unknown default:
+            preconditionFailure("Unsupported session status from BillingServiceNative")
+        }
+    }
+
+    private func tripSku(for sessionType: BillingHandler.SessionType) -> SKUIdentifier {
+        switch sessionType {
+        case .activeGuidance:
+            return .nav2SesTrip
+        case .freeDrive:
+            return .nav2SesFDTrip
+        }
+    }
+}
+
+/**
+ Receives events about navigation changes and triggers appropriate events in `BillingService`.
+
+ Session can be paused (`BillingHandler.pauseBillingSession(with:)`),
+ stopped (`BillingHandler.stopBillingSession(with:)`) or
+ resumed (`BillingHandler.resumeBillingSession(with:)`).
+
+ State of the billing sessions can be obtained using `BillingHandler.sessionState(uuid:)`.
+ */
+final class BillingHandler {
+    /// Parameters on an active session.
+    private struct Session {
+        let type: SessionType
+        /// Indicates whether the session is active but paused.
+        var isPaused: Bool
+    }
+
+    /// The state of the billing session.
+    enum SessionState: Equatable {
+        /// Indicates that there is no active billing session.
+        case stopped
+        /// There is an active paused billing session.
+        case paused
+        /// There is an active running billing session.
+        case running
+    }
+
+    /// Supported session types.
+    enum SessionType: Equatable {
+        case freeDrive
+        case activeGuidance
+
+        var maxSessionInterval: TimeInterval {
+            switch self {
+            case .activeGuidance:
+                return 43200 /*12h*/
+            case .freeDrive:
+                return 3600 /*2h*/
+            }
+        }
+    }
+
+    /// Shared billing handler instance. There is no other instances of `BillingHandler`.
+    private(set) static var shared: BillingHandler = {
+        let accessToken = Directions.shared.credentials.accessToken
```

Comment:
Is it okay that we don't allow setting custom credentials here?

Reply:
@1ec5

Reply:
This allows the developer to set credentials in Info.plist but not at runtime. I suppose that’s unavoidable with something that could potentially be invoked extremely early in the application lifecycle. If a developer needs to set different credentials for specific services, they can still override those at runtime.

Reply:
But we have `NavigationSettings.hared.directions`. Why not using them as it's also a part of the Core?